File: services/api-gateway/app/schemas/appointment.py
import uuid
from datetime import date, datetime, time

from pydantic import BaseModel, Field


# ── Doctor Schedule ──────────────────────────────────────────────────────────


class DoctorScheduleCreate(BaseModel):
    """Schema for creating a doctor schedule slot."""

    doctor_id: uuid.UUID
    department_id: uuid.UUID | None = None
    day_of_week: int = Field(..., ge=0, le=6)  # 0=Mon, 6=Sun
    start_time: time
    end_time: time
    slot_duration_minutes: int = Field(15, ge=5, le=120)
    max_patients: int | None = Field(None, ge=1)
    room: str | None = Field(None, max_length=50)
    consultation_type: str = Field(
        default="general",
        pattern=r"^(general|specialist|follow_up|procedure|anc|dental)$",
    )
    effective_from: date | None = None
    effective_until: date | None = None
    notes: str | None = Field(None, max_length=500)


class DoctorScheduleResponse(BaseModel):
    """Schema for doctor schedule API responses."""

    id: uuid.UUID
    doctor_id: uuid.UUID
    department_id: uuid.UUID | None
    day_of_week: int
    start_time: time
    end_time: time
    slot_duration_minutes: int
    max_patients: int | None
    room: str | None
    consultation_type: str
    is_active: bool
    effective_from: date | None
    effective_until: date | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class DoctorScheduleWithName(DoctorScheduleResponse):
    """Doctor schedule with doctor name for display."""

    doctor_name: str | None = None


# ── Available Slots ──────────────────────────────────────────────────────────


class AvailableSlot(BaseModel):
    """Available time slot for booking."""

    schedule_id: uuid.UUID
    doctor_id: uuid.UUID
    doctor_name: str | None = None
    date: date
    start_time: time
    end_time: time
    room: str | None
    consultation_type: str
    available: bool = True


# ── Appointment ──────────────────────────────────────────────────────────────


class AppointmentCreate(BaseModel):
    """Schema for creating an appointment."""

    patient_id: uuid.UUID
    doctor_id: uuid.UUID
    department_id: uuid.UUID | None = None
    schedule_id: uuid.UUID | None = None
    appointment_date: date
    start_time: time
    end_time: time | None = None
    duration_minutes: int = Field(15, ge=5, le=240)
    appointment_type: str = Field(
        default="consultation",
        pattern=r"^(consultation|follow_up|procedure|lab|radiology|anc|dental|vaccination)$",
    )
    visit_reason: str | None = Field(None, max_length=1000)
    priority: str = Field(
        default="routine", pattern=r"^(routine|urgent|emergency)$"
    )
    room: str | None = Field(None, max_length=50)
    notes: str | None = Field(None, max_length=2000)
    is_recurring: bool = False


class AppointmentResponse(BaseModel):
    """Schema for appointment API responses."""

    id: uuid.UUID
    patient_id: uuid.UUID
    doctor_id: uuid.UUID
    department_id: uuid.UUID | None
    schedule_id: uuid.UUID | None
    encounter_id: uuid.UUID | None
    appointment_number: str
    appointment_date: date
    start_time: time
    end_time: time
    duration_minutes: int
    appointment_type: str
    visit_reason: str | None
    priority: str
    status: str
    checked_in_at: datetime | None
    cancelled_at: datetime | None
    cancellation_reason: str | None
    reminder_sent: bool
    is_recurring: bool
    room: str | None
    notes: str | None
    booked_by: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class AppointmentListItem(BaseModel):
    """Appointment in list view with patient and doctor names."""

    id: uuid.UUID
    appointment_number: str
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    doctor_id: uuid.UUID
    doctor_name: str | None = None
    appointment_date: date
    start_time: time
    end_time: time
    appointment_type: str
    priority: str
    status: str
    room: str | None
    visit_reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AppointmentListResponse(BaseModel):
    """Paginated appointment list response."""

    items: list[AppointmentListItem]
    total: int


class AppointmentUpdate(BaseModel):
    """Schema for updating an appointment."""

    appointment_date: date | None = None
    start_time: time | None = None
    end_time: time | None = None
    status: str | None = Field(
        None,
        pattern=r"^(scheduled|confirmed|checked_in|in_consultation|completed|cancelled|no_show)$",
    )
    room: str | None = Field(None, max_length=50)
    notes: str | None = Field(None, max_length=2000)
    cancellation_reason: str | None = Field(None, max_length=1000)


class AppointmentCheckIn(BaseModel):
    """Schema for checking in a patient."""

    notes: str | None = Field(None, max_length=500)


# ── Summary ──────────────────────────────────────────────────────────────────


class AppointmentSummary(BaseModel):
    """Dashboard summary stats for appointments."""

    total_today: int = 0
    scheduled: int = 0
    confirmed: int = 0
    checked_in: int = 0
    completed_today: int = 0
    cancelled_today: int = 0
    no_show_today: int = 0

File: packages/shared/src/types/appointment.ts
/** Day of the week: 0=Monday, 6=Sunday */
export type DayOfWeek = 0 | 1 | 2 | 3 | 4 | 5 | 6;

/** Consultation type for doctor schedules */
export type ConsultationType =
  | "general"
  | "specialist"
  | "follow_up"
  | "procedure"
  | "anc"
  | "dental";

/** Appointment type */
export type AppointmentType =
  | "consultation"
  | "follow_up"
  | "procedure"
  | "lab"
  | "radiology"
  | "anc"
  | "dental"
  | "vaccination";

/** Appointment priority */
export type AppointmentPriority = "routine" | "urgent" | "emergency";

/** Appointment status lifecycle */
export type AppointmentStatus =
  | "scheduled"
  | "confirmed"
  | "checked_in"
  | "in_consultation"
  | "completed"
  | "cancelled"
  | "no_show";

/** Doctor schedule response */
export interface DoctorScheduleResponse {
  id: string;
  doctor_id: string;
  department_id: string | null;
  day_of_week: DayOfWeek;
  start_time: string;
  end_time: string;
  slot_duration_minutes: number;
  max_patients: number | null;
  room: string | null;
  consultation_type: ConsultationType;
  is_active: boolean;
  effective_from: string | null;
  effective_until: string | null;
  notes: string | null;
  created_at: string;
  doctor_name: string | null;
}

/** Doctor schedule create request */
export interface DoctorScheduleCreate {
  doctor_id: string;
  department_id?: string | null;
  day_of_week: DayOfWeek;
  start_time: string;
  end_time: string;
  slot_duration_minutes?: number;
  max_patients?: number | null;
  room?: string | null;
  consultation_type?: ConsultationType;
  effective_from?: string | null;
  effective_until?: string | null;
  notes?: string | null;
}

/** Available time slot for booking */
export interface AvailableSlot {
  schedule_id: string;
  doctor_id: string;
  doctor_name: string | null;
  date: string;
  start_time: string;
  end_time: string;
  room: string | null;
  consultation_type: ConsultationType;
  available: boolean;
}

/** Appointment create request */
export interface AppointmentCreate {
  patient_id: string;
  doctor_id: string;
  department_id?: string | null;
  schedule_id?: string | null;
  appointment_date: string;
  start_time: string;
  end_time?: string | null;
  duration_minutes?: number;
  appointment_type?: AppointmentType;
  visit_reason?: string | null;
  priority?: AppointmentPriority;
  room?: string | null;
  notes?: string | null;
  is_recurring?: boolean;
}

/** Appointment detail response */
export interface AppointmentResponse {
  id: string;
  patient_id: string;
  doctor_id: string;
  department_id: string | null;
  schedule_id: string | null;
  encounter_id: string | null;
  appointment_number: string;
  appointment_date: string;
  start_time: string;
  end_time: string;
  duration_minutes: number;
  appointment_type: AppointmentType;
  visit_reason: string | null;
  priority: AppointmentPriority;
  status: AppointmentStatus;
  checked_in_at: string | null;
  cancelled_at: string | null;
  cancellation_reason: string | null;
  reminder_sent: boolean;
  is_recurring: boolean;
  room: string | null;
  notes: string | null;
  booked_by: string;
  created_at: string;
}

/** Appointment list item with patient/doctor names */
export interface AppointmentListItem {
  id: string;
  appointment_number: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  doctor_id: string;
  doctor_name: string | null;
  appointment_date: string;
  start_time: string;
  end_time: string;
  appointment_type: AppointmentType;
  priority: AppointmentPriority;
  status: AppointmentStatus;
  room: string | null;
  visit_reason: string | null;
  created_at: string;
}

/** Paginated appointment list response */
export interface AppointmentListResponse {
  items: AppointmentListItem[];
  total: number;
}

/** Appointment update request */
export interface AppointmentUpdate {
  appointment_date?: string | null;
  start_time?: string | null;
  end_time?: string | null;
  status?: AppointmentStatus | null;
  room?: string | null;
  notes?: string | null;
  cancellation_reason?: string | null;
}

/** Check-in request */
export interface AppointmentCheckIn {
  notes?: string | null;
}

/** Dashboard summary stats */
export interface AppointmentSummary {
  total_today: number;
  scheduled: number;
  confirmed: number;
  checked_in: number;
  completed_today: number;
  cancelled_today: number;
  no_show_today: number;
}
